### ISSUE TITLE
How Perfect Match Layer can be more effective？

### ISSUE THREAD

**Issue by Lujiang-ECUST** [ar4/deepwave#86]
Hi, I find that the absorption of the perfectly matched layer is poor when I set 'pml_freq' to the same value as the excitation frequency, even though I set 50 cells all around!What should I do to better absorb the reflected waves at the boundary?
![Figure_1](https://github.com/user-attachments/assets/dc581fb4-2871-4f0d-8f0a-a87f714fdfa6)
The figure shows two distinct boundary reflection waves after propagating for some time

**Comment by Lujiang-ECUST** [comment]
`ny = 200
nx = 200
# dx = 0.5e-4
dx = 0.2e-3
vmin=5800
vmax=6300
Model = md.ModelGenerator('emnist')
model = Model()
# img, label = Eminist_vmodel.EMINST(imaidx, nx, vmin, vmax)
# model = img.reshape((nx, ny))
vp_background = torch.ones(nx,ny, device=device) *5800.0
vs_background = torch.ones(nx,ny, device=device)*3348.6315918
rho_background = torch.ones(nx,ny, device=device)*2.70532036*1000

vp_true = torch.tensor(model['vp'], device=device,dtype=torch.float32)
vs_true = torch.tensor(model['vs'], device=device,dtype=torch.float32)
rho_true = torch.tensor(model['rho']*1000, device=device,dtype=torch.float32)
# vp_true = vp_background.clone()
# vp_true[10:20, 30:40] = 1600
# vs_true = vs_background.clone()
# vs_true[10:20, 45:55] = 1100
# rho_true = rho_background.clone()
# rho_true[10:20, 60:70] = 2300

n_shots = 64

n_sources_per_shot = 1
d_source = 3
first_source = 4
source_depth = 0

n_receivers_per_shot1 = 64
n_receivers_per_shot2 = 32
d_receiver = 3
first_receiver = 4
receiver_depth1 = 0
receiver_depth2 = 99
freq = 2.25e6
nt =1400
# dt = 0.4e-8
dt = 1e-8
peak_time = 1.5 / freq

# source_locations
source_locations = torch.zeros(n_shots, n_sources_per_shot, 2,device=device)
source_locations[..., 0] = source_depth
source_locations[:, 0, 1] = (torch.arange(n_shots) * d_source +
                             first_source)

# receiver_locations
receiver_locations = torch.zeros(n_shots, n_receivers_per_shot1, 2, device=device)
receiver_locations[:,:, 0] = receiver_depth1

receiver_locations[:, :, 1] = (
    (torch.arange(n_receivers_per_shot1) * d_receiver +
     first_receiver)
    .repeat(n_shots, 1)
)



# receiver_locations = torch.zeros(n_shots, n_receivers_per_shot1+ n_receivers_per_shot2, 2, device=device)
# receiver_locations[:,0:n_receivers_per_shot1, 0] = receiver_depth1
# receiver_locations[:,n_receivers_per_shot1:n_receivers_per_shot2+n_receivers_per_shot1, 0] = receiver_depth2
#
# receiver_locations[:, 0:n_receivers_per_shot1, 1] = (
#     (torch.arange(n_receivers_per_shot1) * d_receiver +
#      first_receiver)
#     .repeat(n_shots, 1)
# )
# receiver_locations[:, n_receivers_per_shot1:n_receivers_per_shot2+n_receivers_per_shot1, 1] = (
#     (torch.arange(n_receivers_per_shot2) * d_receiver +
#      first_receiver)
#     .repeat(n_shots, 1)
# )
# source_amplitudes
source_amplitudes = (
    (deepwave.wavelets.ricker(freq, nt, dt, peak_time))
    .repeat(n_shots, n_sources_per_shot, 1).to(device)
)

# Create observed data using true models
# observed_data = elastic(
#     *deepwave.common.vpvsrho_to_lambmubuoyancy(vp_true, vs_true,
#                                                rho_true),
#     dx, dt,
#     accuracy=4,
#     source_amplitudes_y=source_amplitudes,
#     source_locations_y=source_locations,
#     receiver_locations_y=receiver_locations,
#     pml_freq=freq,
# )
observed_data = deepwave.scalar(
                vp_true, dx, dt,
                source_amplitudes=source_amplitudes,
                source_locations=source_locations,
                receiver_locations=receiver_locations,
                max_vel=vp_true.max().max(),
                pml_freq=freq,
                pml_width=[50,50,50,50])
plt.imshow(observed_data[0][0].detach().cpu().numpy())
plt.show()`

**Comment by ar4** [comment]
Hello and thank you for your question.

I ran your code, using the background vp model for the velocity, and I obtained approximately the same image as you did for the final wavefield. However, looking at the amplitudes of the waves, I see that they are extremely small (~1e-12), and so it seems that the PML worked effectively.

Do you see any reflections from the boundaries in your recorded data?

**Comment by ar4** [comment]
I'm going to close this issue now, but please feel free to reopen it or to create a new one if you have any other questions.
